File: ui/src/api/client-networks.js
import { SystemControllerClient } from './core.js'

/**
 * Lists all networks with peer counts and running state.
 *
 * Calls GET /networks on the Control Plane Service.
 *
 * @returns {Promise<Array<{name: string, tld: string, subnet: string, address: string, public_key: string, listen_port: number, enabled: boolean, peer_count: number, interface: string, running: boolean}>>}
 */
SystemControllerClient.prototype.listNetworks = async function () {
  return this.getJSON('/networks')
}

/**
 * Creates a new network. The server derives the overlay subnet from the
 * systemd machine-id and generates a WireGuard keypair.
 *
 * Calls POST /networks/create.
 *
 * @param {string} name - Lowercase alphanumeric-with-dashes network name.
 * @param {string} [tld] - Optional TLD; defaults to the network name.
 * @returns {Promise<object>}
 */
SystemControllerClient.prototype.createNetwork = async function (name, tld) {
  return this.postJSON('/networks/create', { name, tld })
}

/**
 * Removes a network (the default "home" network cannot be removed).
 *
 * Calls POST /networks/remove.
 *
 * @param {string} name
 * @returns {Promise<void>}
 */
SystemControllerClient.prototype.removeNetwork = async function (name) {
  await this.postJSON('/networks/remove', { name })
}

/**
 * Enables a network (brings its WireGuard interface up → remote access on).
 *
 * Calls POST /networks/enable.
 *
 * @param {string} name
 * @returns {Promise<void>}
 */
SystemControllerClient.prototype.enableNetwork = async function (name) {
  await this.postJSON('/networks/enable', { name })
}

/**
 * Disables a network (brings its WireGuard interface down → remote access off;
 * local DNS and services keep running).
 *
 * Calls POST /networks/disable.
 *
 * @param {string} name
 * @returns {Promise<void>}
 */
SystemControllerClient.prototype.disableNetwork = async function (name) {
  await this.postJSON('/networks/disable', { name })
}

/**
 * Lists the peers on a network.
 *
 * Calls GET /networks/peers?network=<name>.
 *
 * @param {string} network
 * @returns {Promise<Array<{network: string, public_key: string, name: string, allowed_ip: string, endpoint: string}>>}
 */
SystemControllerClient.prototype.listNetworkPeers = async function (network) {
  return this.getJSON(`/networks/peers?network=${encodeURIComponent(network)}`)
}

/**
 * Adds a peer to a network. When public_key is omitted the server generates a
 * keypair and returns the private key plus a ready-to-import device config.
 *
 * Calls POST /networks/peers/add.
 *
 * @param {string} network
 * @param {string} name - Human-readable device label.
 * @param {string} [publicKey] - The device's own public key (optional).
 * @param {string} [endpoint] - The device's reachable endpoint (optional).
 * @returns {Promise<{peer: object, private_key?: string, config: string}>}
 */
SystemControllerClient.prototype.addNetworkPeer = async function (network, name, publicKey, endpoint) {
  return this.postJSON('/networks/peers/add', { network, name, public_key: publicKey, endpoint })
}

/**
 * Removes a peer from a network.
 *
 * Calls POST /networks/peers/remove.
 *
 * @param {string} network
 * @param {string} publicKey
 * @returns {Promise<void>}
 */
SystemControllerClient.prototype.removeNetworkPeer = async function (network, publicKey) {
  await this.postJSON('/networks/peers/remove', { network, public_key: publicKey })
}

File: src/svc/systemcontroller/client_networks.go
package systemcontroller

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"

	"gitea.com/town-os/town-os/src/account"
)

// ListNetworks returns all overlay networks with peer counts and running state.
func (c *SystemdClient) ListNetworks(ctx context.Context) (_ []NetworkView, err error) {
	resp, err := c.getClient(ctx, "networks")
	if err != nil {
		return nil, fmt.Errorf("%w: ListNetworks: %w", ErrHTTPRequest, err)
	}
	defer func() { err = errors.Join(err, resp.Body.Close()) }()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "GET", "networks")
	}

	var views []NetworkView
	return views, json.NewDecoder(resp.Body).Decode(&views)
}

// CreateNetwork creates an overlay network with the given name and optional TLD
// (empty TLD defaults to the name server-side) and returns the created network.
func (c *SystemdClient) CreateNetwork(ctx context.Context, name, tld string) (_ NetworkView, err error) {
	pr, pw := io.Pipe()
	go pipeEncode(pw, CreateNetworkRequest{Name: name, TLD: tld})

	resp, err := c.postJSON(ctx, "networks/create", pr)
	if err != nil {
		return NetworkView{}, fmt.Errorf("%w: CreateNetwork: %w", ErrHTTPRequest, err)
	}
	defer func() { err = errors.Join(err, resp.Body.Close()) }()

	if resp.StatusCode != http.StatusOK {
		return NetworkView{}, readProblemDetail(resp, "POST", "networks/create")
	}

	var view NetworkView
	return view, json.NewDecoder(resp.Body).Decode(&view)
}

// RemoveNetwork deletes a network by name. The default network cannot be removed.
func (c *SystemdClient) RemoveNetwork(ctx context.Context, name string) error {
	pr, pw := io.Pipe()
	go pipeEncode(pw, NetworkNameRequest{Name: name})
	return c.postClient(ctx, "networks/remove", pr)
}

// EnableNetwork brings a network's overlay interface up (remote access on).
func (c *SystemdClient) EnableNetwork(ctx context.Context, name string) error {
	pr, pw := io.Pipe()
	go pipeEncode(pw, NetworkNameRequest{Name: name})
	return c.postClient(ctx, "networks/enable", pr)
}

// DisableNetwork brings a network's overlay interface down (remote access off;
// local services keep working).
func (c *SystemdClient) DisableNetwork(ctx context.Context, name string) error {
	pr, pw := io.Pipe()
	go pipeEncode(pw, NetworkNameRequest{Name: name})
	return c.postClient(ctx, "networks/disable", pr)
}

// ListNetworkPeers returns the peers registered on the named network.
func (c *SystemdClient) ListNetworkPeers(ctx context.Context, network string) (_ []account.NetworkPeer, err error) {
	path := "networks/peers?network=" + url.QueryEscape(network)
	resp, err := c.getClient(ctx, path)
	if err != nil {
		return nil, fmt.Errorf("%w: ListNetworkPeers: %w", ErrHTTPRequest, err)
	}
	defer func() { err = errors.Join(err, resp.Body.Close()) }()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "GET", "networks/peers")
	}

	var peers []account.NetworkPeer
	return peers, json.NewDecoder(resp.Body).Decode(&peers)
}

// AddNetworkPeer registers a peer on a network. When PublicKey is empty the
// server generates a keypair and returns the private key plus a device config.
func (c *SystemdClient) AddNetworkPeer(ctx context.Context, req AddNetworkPeerRequest) (_ AddPeerResult, err error) {
	pr, pw := io.Pipe()
	go pipeEncode(pw, req)

	resp, err := c.postJSON(ctx, "networks/peers/add", pr)
	if err != nil {
		return AddPeerResult{}, fmt.Errorf("%w: AddNetworkPeer: %w", ErrHTTPRequest, err)
	}
	defer func() { err = errors.Join(err, resp.Body.Close()) }()

	if resp.StatusCode != http.StatusOK {
		return AddPeerResult{}, readProblemDetail(resp, "POST", "networks/peers/add")
	}

	var result AddPeerResult
	return result, json.NewDecoder(resp.Body).Decode(&result)
}

// RemoveNetworkPeer removes a peer from a network by its public key.
func (c *SystemdClient) RemoveNetworkPeer(ctx context.Context, network, publicKey string) error {
	pr, pw := io.Pipe()
	go pipeEncode(pw, RemoveNetworkPeerRequest{Network: network, PublicKey: publicKey})
	return c.postClient(ctx, "networks/peers/remove", pr)
}

// RefreshNetworkPeer extends a peer's TTL by the server's configured peer_ttl
// and returns the new expiry. This is the heartbeat a long-lived client issues
// to keep its enrollment from being reaped.
func (c *SystemdClient) RefreshNetworkPeer(ctx context.Context, network, publicKey string) (_ RefreshPeerResult, err error) {
	pr, pw := io.Pipe()
	go pipeEncode(pw, RefreshNetworkPeerRequest{Network: network, PublicKey: publicKey})

	resp, err := c.postJSON(ctx, "networks/peers/refresh", pr)
	if err != nil {
		return RefreshPeerResult{}, fmt.Errorf("%w: RefreshNetworkPeer: %w", ErrHTTPRequest, err)
	}
	defer func() { err = errors.Join(err, resp.Body.Close()) }()

	if resp.StatusCode != http.StatusOK {
		return RefreshPeerResult{}, readProblemDetail(resp, "POST", "networks/peers/refresh")
	}

	var result RefreshPeerResult
	return result, json.NewDecoder(resp.Body).Decode(&result)
}
